### PR TITLE
(enh) Improve CodeActAgent's file editing reliability

### PR DESCRIPTION
**Pull request by tobitege**
**Short description of the problem this fixes or functionality that this introduces. This may be used for the CHANGELOG**

Improve system prompt to reduce file editing errors made by LLM.

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

A recurring issue with the agent skills for file editing and models like Sonnet were incorrect calls of methods, that caused extra steps for the LLM to get it right.

Issues like:
- First attempt to create a new file was followed by a replace command with empty content, which we do not allow. Causing multiple, preventable steps and costs.
- An `open_file` only shows a limited number of lines and it is apparently _really hard_ to convince an LLM to read a file fully in order to analyze it. This often resulted in Sonnet to just _assume_ to know a file content from just the initial 100 lines.
Subsequent edits of it to the same file then resulted in replacement not found (no edit action performed -> edit loop), or the file content becoming incomplete and/or syntactically wrong.

Changes (in short):
- The system prompt (`system_prompt.j2`) text and file ops docstrings got applied several tweaks with concrete hints.
- `file_ops.py`: in order to reduce noise in the logs, errors are mostly output as plain text to the LLM and no longer by raising an exception (avoiding lengthy tracebacks)
- Unit tests adopted accordingly
- Integration tests: removed references to server runtime (incl. mock .log files) and TEST_RUNTIME

**TODO:**
- [X] Units tests are working
- [x] Integration tests need regeneration
- [ ] SWE bench eval

---
This is a rebased PR version of #3550 but now on repo branch, so other maintainers can edit.

### CODE REVIEW

**Comment by amanape** [comment]
Is there a way we can outsource the content in the test prompt logs to one place so that it is easier to edit and iterate on 🤔? Editing each individual file seems like a pain, especially if and when we will want to run various evaluations on different kind of prompts

File: openhands/runtime/plugins/agent_skills/file_ops/file_ops.py
Comment (xingyaoww) [inline openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:128]: nice improvement!!

Comment (xingyaoww) [reply openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:128]: But, if we are scrolling up, shouldn't the CURRENT_LINE be the last line? if that make sense?

File: openhands/runtime/plugins/agent_skills/file_ops/file_ops.py
Comment (xingyaoww) [inline openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:560]: nit: We could make a global variable `SEP` so we can re-use it in other function, but it is purely a style choice -- take it with a grain of salt

File: tests/integration/mock/server_runtime/CodeActAgent/test_browse_internet/prompt_001.log
Comment (xingyaoww) [inline tests/integration/mock/server_runtime/CodeActAgent/test_browse_internet/prompt_001.log:1]: It occured to me: maybe we can remove all the mock tests for server runtime, since we already deprecated it?

File: openhands/runtime/plugins/agent_skills/file_ops/file_ops.py
Comment (xingyaoww) [inline openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:1]: These are great improvements! Especially about the window - it actually make me think about more stuff we can improve the window:

1. When open each file, we could use parser to parse all the elements (e.g., all the function names, class names, and their corresponding line number - similar to https://github.com/ozyyshr/RepoGraph) and print it on the top. For example:
```
[Structure of the file]
LINE 5| def balabal(a: int, b:str): 
...
LINE 10| class Hello:
LINE 10|     def world(self, haha: str):
```
This could make things easier for the agent to navigate the file (e.g., can directly *go to* specific line if it is just trying to look for some file).

**Review by xingyaoww — COMMENTED** [review]
This is amazing! Just had a tiny question about the window behavior when scroll up. I can help regenerate the integration tests, and run SWE-Bench when this is ready.

Once we have AiderBench results and SWE-Bench result, i think this is probably ready to merge!

File: openhands/runtime/plugins/agent_skills/file_ops/file_ops.py
Comment (tobitege) [reply openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:128]: Oh, good catch, I'll look into this asap!

File: openhands/runtime/plugins/agent_skills/file_ops/file_ops.py
Comment (tobitege) [reply openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:1]: I'd leave that to when we get something like a repo map.

File: openhands/runtime/plugins/agent_skills/file_ops/file_ops.py
Comment (tobitege) [reply openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:560]: I didn't find much more use of that except within the `test_agent_skill.py`

**Comment by tobitege** [comment]
Welp, that's like 23K lines of mock files gone! 🤣 
@xingyaoww this should be it for edits for now

File: tests/integration/regenerate.sh
Comment (xingyaoww) [inline tests/integration/regenerate.sh:187]: I think we should still keep this and make this configurable? Bc we may add different runtime in the future?

File: tests/integration/test_agent.py
Comment (xingyaoww) [inline tests/integration/test_agent.py:20]: ditto - i think we shouldn't hard code it?

File: tests/runtime/conftest.py
Comment (xingyaoww) [inline tests/runtime/conftest.py:38]: ditto

**Review by xingyaoww — COMMENTED** [review]
I think we maybe need to revert these changes related to "TEST_RUNTIME"? other than that i think this is good - will regen integration tests now

File: .github/workflows/ghcr_runtime.yml
Comment (xingyaoww) [inline .github/workflows/ghcr_runtime.yml:195]: ditto

File: tests/integration/regenerate.sh
Comment (tobitege) [reply tests/integration/regenerate.sh:187]: I don't understand why we would keep this dead code now when we try to clean up stuff in other areas? 🤔

Comment (xingyaoww) [reply tests/integration/regenerate.sh:187]: I'm trying to implement a different RemoteRuntime in a different PR, and when that finishes I likely want to add it to the runtime tests and integration test. Keeping this TEST_RUNTIME argument will make it easier for us to extend the tests to different runtime types in the future

Comment (tobitege) [reply tests/integration/regenerate.sh:187]: Ahh, ok, got it. Will do the reverts then.

Comment (xingyaoww) [reply tests/integration/regenerate.sh:187]: Thanks for all the hassle 😃

Comment (tobitege) [reply tests/integration/regenerate.sh]: All good, done now. 😄

File: tests/integration/test_agent.py
Comment (tobitege) [reply tests/integration/test_agent.py:20]: Resolved

**Comment by xingyaoww** [comment]
I started running eval on the latest commit - will report back when it finishes

File: agenthub/codeact_agent/system_prompt.j2
Comment (xingyaoww) [inline agenthub/codeact_agent/system_prompt.j2]: New prompt changes after the evaluation started 😩 - Can we hold off on these prompt changes to a separate PR so we can ensure the number I'm about to finish evaluating accurately reflects the prompt for this PR?

Comment (tobitege) [reply agenthub/codeact_agent/system_prompt.j2]: Doh! I didn't think of that, sorry! Will revert it asap! 😔

Comment (xingyaoww) [reply agenthub/codeact_agent/system_prompt.j2]: No worries at all! 😄  just making sure our number accurately reflect the state of this PR

**Comment by tobitege** [comment]
DO NOT MERGE until Ok from Xingyaoww (after bench result)

**Comment by xingyaoww** [comment]
89/300 - definitely some improvements from the previous main (i remember that was 84-85 ish). Will upload trajectories soon. 
image

**Review by xingyaoww — APPROVED** [review]

**Comment by tobitege** [comment]
> 89/300 - definitely some improvements from the previous main (i remember that was 84-85 ish). Will upload trajectories soon. Let's get this in!

Thanks so much for putting it through the bench, most people don't know how much time and observations it takes! 🤗

**Pull request merged**
